Code/Framework/AzCore/AzCore/EBus/EBus.h

```diff
@@ -619,12 +636,42 @@ namespace AZ
              */
             using ContextMutexType = AZStd::conditional_t<BusTraits::LocklessDispatch && AZStd::is_same_v<MutexType, AZ::NullMutex>, AZStd::shared_mutex, MutexType>;
 
+            struct ThreadPostDispatchLockGuard
+            {
+                struct DispatchPolicyInvoker
+                {
+                    DispatchPolicyInvoker(Context& context)
+                        : m_context{ context }
+                    {}
+                    ~DispatchPolicyInvoker()
+                    {
+                        if constexpr (EnablePostDispatch)
+                        {
+                            if (!IsInDispatchThisThread(&m_context))
+                            {
+                                typename Traits::ThreadDispatchPolicy{}(EBusPolicies::PostDispatchTag);
+                            }
+                        }
+                    }
+
+                    Context& m_context;
+                };
+                ThreadPostDispatchLockGuard(Context& context, ContextMutexType& contextMutex);
+                ThreadPostDispatchLockGuard(Context& context, ContextMutexType& contextMutex, AZStd::adopt_lock_t);
+                ThreadPostDispatchLockGuard(const ThreadPostDispatchLockGuard&) = delete;
+                ~ThreadPostDispatchLockGuard() = default;
+                ThreadPostDispatchLockGuard& operator=(const ThreadPostDispatchLockGuard&) = delete;
+            private:
+                DispatchPolicyInvoker m_threadPolicyInvoker;
+                using LockType = AZStd::conditional_t<BusTraits::LocklessDispatch, AZ::Internal::NullLockGuard<ContextMutexType>, AZStd::scoped_lock<ContextMutexType>>;
+                LockType m_lock;
+            };
             /**
              * The scoped lock guard to use (either AZStd::scoped_lock<MutexType> or NullLockGuard<MutexType>
              * during broadcast/event dispatch.
              * @see EBusTraits::LocklessDispatch
              */
-            using DispatchLockGuard = AZStd::conditional_t<BusTraits::LocklessDispatch, AZ::Internal::NullLockGuard<ContextMutexType>, AZStd::scoped_lock<ContextMutexType>>;
+            using DispatchLockGuard = ThreadPostDispatchLockGuard;
```

Comment:
Just wondering if this entire mechanic needs to be part of the EBus itself. Would be be possible to instead allow the EBus to optionally take an instance of a specific lock instead of creating its own and then keep this entire mechanic outside the EBus through a special kind of lock with the additional requirement (which `ThreadPostDispatchLockGuard` seems to already be moving to)?

Reply:
That sounds good. We can move this mechanic outside of the EBus Context itself.
We can then allow the user to supply the EBusTraits with a DispatchLockType itself.

Reply:
The EBus code has been updated to allow the user to supply a DispatchLockGuard alias template, that can be used to customize the behavior of Locking/Unlocking during Dispatch.

This allows me to replicate the same behavior as before, but with only minor modifications to the EBus code.